Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/non_rigid_mesh_registration.h

```diff
@@ -0,0 +1,758 @@
+      dist(i, j) = it->second;
+    }
+  }
+
+  return std::make_pair(idz, dist);
+}
+
+template <typename T>
+int sign(T val) {
+  return (T(0) < val) - (val < T(0));
+}
+
+void insertSparseMatrix(const SparseMat& mat, std::vector<SparseTriplet>& coefficients, size_t start_i = 0, size_t start_j = 0) {
+  for (int k = 0; k < mat.outerSize(); ++k)
+    for (SparseMat::InnerIterator it(mat, k); it; ++it)
+      coefficients.push_back(SparseTriplet(start_i + it.row(), start_j + it.col(), it.value()));
+}
+
+template <typename T>
+Eigen::Matrix<T, 3, 3> rot(T a, T b, T c) {
+  T ca = cos(a), cb = cos(b), cc = cos(c);
+  T sa = sin(a), sb = sin(b), sc = sin(c);
+  Eigen::Matrix<T, 3, 3> R;
+  R << cb * cc, cc* sa* sb - ca * sc, ca* cc* sb + sa * sc,
+    cb* sc, ca* cc + sa * sb * sc, ca* sb* sc - cc * sa,
+    -sb, cb* sa, ca* cb;
+  return R;
+}
+
+} // namespace internal
+
+/*!
+* \ingroup PMP_registration_grp
+*
+* \brief calculates non-rigid transformation of a mesh onto a set of oriented points.
+*
+* A non-rigid ICP, iterative closest point, method based on
+* <A HREF="https://vgl.ict.usc.edu/Research/NonRigidRegistration/MODERN%20TECHNIQUES%20AND%20APPLICATIONS%20FOR%20REAL-TIME%20NON-RIGID%20REGISTRATION.pdf">a SIGGRAPH'16 Tutorial</A>.
+* The method uses a few `correspondences` between the `source` and the `target` for the rough alignment. The iterative closest point method
+* iteratively approaches the `target` by minimizing the distance between vertices of the `source` and points of the `target`.
+*
+* @note This function requires the \ref thirdpartyEigen library.
+*
+* @tparam TriangleMesh a model of `MutableFaceGraph`.
+* @tparam PointRange a model of the `concept ForwardRange` whose value type is the point type.
+* @tparam VertexTranslationMap is a property map with `boost::graph_traits<TriangleMesh1>::%vertex_descriptor`
+*   as key type and a \cgal Kernel `Vector_3` as value type.
+* @tparam VertexRotationMap is a property map with `boost::graph_traits<TriangleMesh1>::%vertex_descriptor`
+*   as key type and a \cgal Kernel `Aff_transformation_3` as value type.
+* @tparam NamedParameters1 a sequence of \ref bgl_namedparameters "Named Parameters"
+* @tparam NamedParameters2 a sequence of \ref bgl_namedparameters "Named Parameters"
+*
+* @param source the triangle mesh that should be mapped onto `target`.
+* @param target the target triangle mesh.
+* @param vtm a writeable vertex property map of `source` to store the translation vector of the registration.
+* @param vrm a writeable vertex property map of `source` to store the rotation part of the registration.
+* @param np1 an optional sequence of \ref bgl_namedparameters "Named Parameters" among the ones listed below
+* @param np2 an optional sequence of \ref bgl_namedparameters "Named Parameters" providing a point_map and normal_map for the `PointRange`
+* @param correspondences a vector given matching points between the `source` and the `target`
+*
+* \cgalNamedParamsBegin
```

Comment:
All these numbers are opaque. What is a reasonable range?  Why the defaults? Do they have any dependence on dimensions of the input  (average edge length, bbox,....) ?

Reply:
The weights have to be found by trial and error. They are weights of the energy terms, so I added that they need to be >= 0. I added a reference the the user manual.